src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -8,393 +8,99 @@
-			*randomcoeff_multiplicative2);
-
-
-}
-
-void RandomFourierGaussPreproc::set_randomcoefficients(
-		float64_t *randomcoeff_additive2,
-		float64_t * randomcoeff_multiplicative2,
-		const int32_t dim_feature_space2, const int32_t dim_input_space2, const float64_t kernelwidth2) {
-	dim_feature_space = dim_feature_space2;
-	dim_input_space = dim_input_space2;
-	kernelwidth=kernelwidth2;
-
-	SG_FREE(randomcoeff_multiplicative);
-	randomcoeff_multiplicative=NULL;
-	SG_FREE(randomcoeff_additive);
-	randomcoeff_additive=NULL;
-
-	cur_dim_feature_space=dim_feature_space;
-	cur_dim_input_space = dim_input_space;
-	cur_kernelwidth=kernelwidth;
-
-	if( (dim_feature_space>0) && (dim_input_space>0) )
-	{
-	randomcoeff_additive=SG_MALLOC(float64_t, cur_dim_feature_space);
-	randomcoeff_multiplicative=SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	std::copy(randomcoeff_additive2, randomcoeff_additive2
-			+ dim_feature_space, randomcoeff_additive);
-	std::copy(randomcoeff_multiplicative2, randomcoeff_multiplicative2
-			+ cur_dim_feature_space*cur_dim_input_space, randomcoeff_multiplicative);
-	}
+	UniformRealDistribution<float64_t> uniform(0.0, 2.0 * Math::PI);
 
+	for (auto row : range(m_dim_output))
+		m_offset[row] = uniform(m_prng);
 }
 
 void RandomFourierGaussPreproc::fit(std::shared_ptr<Features> f)
 {
-	if (dim_feature_space <= 0) {
-		throw ShogunException(
-				"CRandomFourierGaussPreproc::init (Features *f): dim_feature_space<=0 is not allowed, use void set_dim_feature_space(const int32 dim) before!\n");
-	}
-
-	io::info("calling CRandomFourierGaussPreproc::init(...)");
-	int32_t num_features =
-	    f->as<DenseFeatures<float64_t>>()->get_num_features();
-
-	if (!test_rfinited()) {
-		dim_input_space = num_features;
-		init_randomcoefficients();
-		ASSERT( test_rfinited())
-	} else {
-		dim_input_space = num_features;
-		// does not reinit if dimension is the same to avoid overriding a previous call of set_randomcoefficients(...)
-		init_randomcoefficients();
-	}
+	auto num_features = f->as<DenseFeatures<float64_t>>()->get_num_features();
+	require(num_features > 0, "Input space dimension must be greater than zero");
+	
+	init_basis(num_features);
+	m_fitted = true;
 }
 
 SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(SGVector<float64_t> vector)
 {
-	if (!test_rfinited()) {
-		throw ShogunException(
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
+	SGVector<float64_t> projection(m_dim_output);
 
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
+	assert_fitted();
+	linalg::matrix_prod(m_basis, vector, projection);
+	linalg::add(projection, m_offset, projection);
```

Comment:
Ok one more thing. Sorry!
But this code simply replicates the matrix code. Why have it twice? It is pretty much copy pasted ....

 You can just call the matrix code from here. All you need to do is to make a matrix from the input vector
`return apply_to_feature_matrix(SGMatrix<...>(vector)).get_column_vector(0)` (double check the details ... )

Reply:
you would have to make the column vector persistent somehow as well

Reply:
Done!